js/src/Figure.js

```diff
@@ -741,7 +741,25 @@ var FigureView = widgets.DOMWidgetView.extend({
         var initialTraces = _.cloneDeep(this.model.get("_data"));
         var initialLayout = _.cloneDeep(this.model.get("_layout"));
 
-        Plotly.newPlot(that.el, initialTraces, initialLayout).then(
+        let holder = document.createElement('div');
+        holder.classList.add('js-plotly-plot-inner');
+        holder.style.height = '100%';
+        holder.style.width = '100%';
+
+        that.plot = holder;
+
+        that.el.appendChild(holder);
+        that.el.style.resize = 'both';
+        that.el.style.paddingBottom = '10px';
+
+        let observer = new MutationObserver(that.autosizeFigure.bind(that));
+        observer.observe(this.el, {attributes: true});
+
+        that.el.addEventListener('resize', function(){
+            that.autosizeFigure();
+        });
```

Comment:
Could you explain why the `MutationObserver` and `addEventListener` approaches are both needed here? And add the explanations as comments 🙂

Reply:
Mutationobserver monitors for changes in the element, while resize events are only thrown for window size changes (not for element size changes). When the window resizes, the element can just overflow (which we don't want) so we want both the element and the window to trigger resizing if they change size